### PR TITLE
[DAL] Data management: table-related type APIs described

### PR DESCRIPTION
**Pull request by michael-smirnov**
Table — oneDAL documentation -
Homogeneous table — oneDAL documentation -

### CODE REVIEW

File: source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    template <typename Data>
+    static homogen_table wrap(const sycl::queue& queue,
+                              const Data* data_pointer,
+                              std::int64_t row_count,
+                              std::int64_t column_count,
+                              const sycl::vector_class<sycl::event>& dependencies = {},
+                              data_layout layout = data_layout::row_major);
+
+public:
+    /// Creates new ``homogen_table`` instance with zero row and column count.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp:35]: "Creates new" -> "Creates a new" here and other places

File: source/elements/oneDAL/source/data_management/table.rst [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+   data_layout::column_major
+      The data block elements are stored column-by-column.
+
+.. _feature_type:
+
+Feature type
+------------
+
+An implementation of the data logical types.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/table.rst:146]: logical data types?

File: source/elements/oneDAL/source/data_management/table.rst [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+be :term:`reference-counted <Reference-counted object>`.
+
+.. onedal_class:: oneapi::dal::table_metadata
+
+.. _data_layout:
+
+Data layout
+-----------
+
+An implementation of the :capterm:`data layout` term.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/table.rst:123]: an implementation of term? concept maybe?

File: source/elements/oneDAL/source/data_management/table.rst [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+.. list-table::
+   :header-rows: 1
+   :widths: 30 70
+
+   * - Table type
+     - Description
+
+   * - :txtref:`homogen_table`
+     - Dense table that contains :term:`contiguous <Contiguous data>`
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/table.rst:88]:
```suggestion
     - A dense table that contains :term:`contiguous <Contiguous data>`
```

File: source/elements/oneDAL/source/data_management/table.rst [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+2. Be derived from the :expr:`oneapi::dal::table` class. The behavior of this class can be
+   extended, but cannot be weaken.
+
+3. Be :term:`reference-counted <Reference-counted object>`.
+
+4. Every new :expr:`oneapi::dal::table` sub-type shall define a unique id number - the "kind"
+   that represents objects of that type in runtime.
+
+The following listing provides an example of table API regarding kinds and copy-assignment operation:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/table.rst:50]:
```suggestion
The following listing provides an example of table API to illustrate table kinds and copy-assignment operation:
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    /// Creates new table instance and moves all property values from $other$ into it.
+    table(table&& other);
+
+    /// Replaces the values of all properties by references to property values of $other$
+    table& operator=(const table& other);
+
+    /// Swaps the property values of this object and $other$.
+    table& operator=(table&& other);
+
+    /// Indicates whether table contains non-zero count of rows and columns or not
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:51]:
```suggestion
    /// Indicates whether a table contains non-zero count of rows and columns or not.
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    bool has_data() const noexcept;
+
+    /// The count of columns in the table.
+    /// @remark default = 0
+    std::int64_t get_column_count() const;
+
+    /// The count of rows in the table.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:58]:
```suggestion
    /// The number of columns in the table.
    /// @remark default = 0
    std::int64_t get_column_count() const;

    /// The number of rows in the table.
```

File: source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    /// The :expr:`data` shall point to the $data_pointer$ memory block.
+    ///
+    /// @tparam Data        The type of elements in the data block that will be stored into the table.
+    ///                     The table shall initialize data types of metadata with this data type.
+    ///                     The feature types shall be set to default values for $Data$ type: contiguous for floating-point,
+    ///                     ordinal for integer types.
+    /// @param queue        The SYCL* queue object
+    /// @param data_pointer The pointer to homogeneous data block.
+    /// @param row_count    The count of rows in the table.
+    /// @param column_count The count of columns in the table.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp:22]:
```suggestion
    /// @param row_count    The number of rows in the table.
    /// @param column_count The number of columns in the table.
```

File: source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    /// The :expr:`data` shall point to the $data_pointer$ memory block.
+    ///
+    /// @tparam Data         The type of elements in the data block that will be stored into the table.
+    /// @tparam ConstDeleter The type of deleter called on $data_pointer$ when
+    ///                      the last table that refers to it goes out of the scope.
+    ///
+    /// @param queue         The SYCL* queue object
+    /// @param data_pointer  The pointer to homogeneous data block.
+    /// @param row_count     The count of rows in the table.
+    /// @param column_count  The count of columns in the table.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp:50]:
```suggestion
    /// @param row_count     The number of rows in the table.
    /// @param column_count  The number of columns in the table.
```

File: source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    /// Creates new ``homogen_table`` instance from externally-defined data block.
+    /// Table object owns the data pointer.
+    /// The :expr:`data` shall point to the $data_pointer$ memory block.
+    ///
+    /// @tparam Data         The type of elements in the data block that will be stored into the table.
+    /// @tparam ConstDeleter The type of deleter called on $data_pointer$ when
+    ///                      the last table that refers to it goes out of the scope.
+    ///
+    /// @param queue         The SYCL* queue object
+    /// @param data_pointer  The pointer to homogeneous data block.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp:48]:
```suggestion
    /// @param data_pointer  The pointer to a homogeneous data block.
```

File: source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    /// object refers to the data but does not own it. The responsibility to
+    /// free the data remains on the user side.
+    /// The :expr:`data` shall point to the $data_pointer$ memory block.
+    ///
+    /// @tparam Data        The type of elements in the data block that will be stored into the table.
+    ///                     The table shall initialize data types of metadata with this data type.
+    ///                     The feature types shall be set to default values for $Data$ type: contiguous for floating-point,
+    ///                     ordinal for integer types.
+    /// @param queue        The SYCL* queue object
+    /// @param data_pointer The pointer to homogeneous data block.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/homogen.hpp:20]:
```suggestion
    /// @param data_pointer The pointer to a homogeneous data block.
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+enum class feature_type { nominal, ordinal, interval, ratio };
+
+enum class data_layout { unknown, row_major, column_major };
+
+class table_metadata {
+public:
+    /// Creates the metadata instance without information about features.
+    /// The :expr:`feature_count` shall be set to zero.
+    /// The properties :expr:`data_type` and :expr:`feature_type` shall not be initialized.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:11]:
```suggestion
    /// The :expr:`data_type` and :expr:`feature_type` properties shall not be initialized.
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
+namespace oneapi::dal {
+
+enum class feature_type { nominal, ordinal, interval, ratio };
+
+enum class data_layout { unknown, row_major, column_major };
+
+class table_metadata {
+public:
+    /// Creates the metadata instance without information about features.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:9]:
```suggestion
    /// Creates the metadata instance without information about the features.
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    /// The :expr:`feature_count` shall be set to zero.
+    /// The properties :expr:`data_type` and :expr:`feature_type` shall not be initialized.
+    table_metadata();
+
+    /// Creates the metadata instance from external information about data types and
+    /// feature types.
+    /// @param dtypes Data types of the features. Shall be assigned into the :expr:`data_type` property.
+    /// @param ftypes Feature types of features. Shall be assigned into the :expr:`feature_type` property.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:17]:
```suggestion
    /// Creates the metadata instance from external information about the data types and
    /// the feature types.
    /// @param dtypes The data types of the features. Shall be assigned into the :expr:`data_type` property.
    /// @param ftypes The feature types of the features. Shall be assigned into the :expr:`feature_type` property.
```

Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:17]:
```suggestion
    /// Creates the metadata instance from external information about the data types and
    /// the feature types.
    /// @param dtypes The data types of the features. Shall be assigned into the :expr:`data_type` property.
    /// @param ftypes The feature types of the features. Shall be assigned into the :expr:`feature_type` property.
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    /// @pre :expr:`dtypes.get_count() == ftypes.get_count()`
+    table_metadata(const array<data_type>& dtypes, const array<feature_type>& ftypes);
+
+    /// The number of features that metadata has information about
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:21]:
```suggestion
    /// The number of features that metadata contains information about
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    /// @pre :expr:`feature_count >= 0`
+    std::int64_t get_feature_count() const;
+
+    /// Feature types in the metadata object. Shall be in range ``[0, feature_count)``
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:25]:
```suggestion
    /// Feature types in the metadata object. Shall be within the range ``[0, feature_count)``
```

File: source/elements/oneDAL/include/oneapi/dal/table/common.hpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    const feature_type& get_feature_type(std::int64_t feature_index) const;
+
+    /// Data types of the features in the metadata object. Shall be in range ``[0, feature_count)``
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/table/common.hpp:28]:
```suggestion
    /// Data types of the features in the metadata object. Shall be within the range ``[0, feature_count)``
```

**Review by andreynv — APPROVED** [review]

**Pull request merged**